src/js_subscriptions.hpp

```diff
@@ -0,0 +1,815 @@
+    return_value.set((uint32_t)subs->version());
+}
+
+/**
+ * @brief Get an array snapshot of the subscriptions in the subscription set
+ *
+ * @param ctx JS context
+ * @param object \ref ObjectType wrapping the SubscriptionSet
+ * @param args \ref None
+ * @param return_value \ref ReturnValue wrapping an array of Subscription instances
+ */
+template <typename T>
+void SubscriptionsClass<T>::snapshot(ContextType ctx, ObjectType this_object, Arguments& args,
+                                     ReturnValue& return_value)
+{
+    args.validate_count(0);
+
+    auto subs = get_internal<T, SubscriptionsClass<T>>(ctx, this_object);
+
+    auto js_subs = std::vector<ValueType>();
+    for (auto& sub : *subs) {
+        js_subs.emplace_back(SubscriptionClass<T>::create_instance(ctx, sub));
+    }
+
+    auto subs_array = Object::create_array(ctx, js_subs);
+    return_value.set(subs_array);
+}
+
+/**
+ * @brief Find a subscription by name
+ *
+ * @param ctx JS context
+ * @param object \ref ObjectType wrapping the SubscriptionSet
+ * @param args \ref A single argument containing the subscription name to find
+ * @param return_value \ref ReturnValue wrapping a Subscription if found, null if not
+ */
+template <typename T>
+void SubscriptionsClass<T>::find_by_name(ContextType ctx, ObjectType this_object, Arguments& args,
+                                         ReturnValue& return_value)
+{
+    args.validate_count(1);
+
+    std::string name = Value::validated_to_string(ctx, args[0], "name");
+    auto subs = get_internal<T, SubscriptionsClass<T>>(ctx, this_object);
+
+    auto it = subs->find(name);
+
+    if (it != subs->end()) {
+        auto sub = SubscriptionClass<T>::create_instance(ctx, *it);
+        return_value.set(sub);
+    }
+    else {
+        return_value.set_null();
+    }
+}
+
+/**
+ * @brief Find a subscription by query
+ *
+ * @param ctx JS context
+ * @param object \ref ObjectType wrapping the SubscriptionSet
+ * @param args \ref A single argument containing the query to find, represented as a Results instance
+ * @param return_value \ref ReturnValue wrapping a Subscription if found, null if not
+ * @exception std::runtime_error if the argument is not a Results instance
+ */
+template <typename T>
+void SubscriptionsClass<T>::find(ContextType ctx, ObjectType this_object, Arguments& args, ReturnValue& return_value)
+{
+    args.validate_count(1);
+
+    auto results_arg = Value::validated_to_object(ctx, args[0], "query");
+    if (!Object::template is_instance<ResultsClass<T>>(ctx, results_arg)) {
+        throw std::runtime_error("Argument to 'findByName' must be a collection of Realm objects.");
+    }
+
+    auto subs = get_internal<T, SubscriptionsClass<T>>(ctx, this_object);
+    auto results = get_internal<T, ResultsClass<T>>(ctx, results_arg);
+    auto query = results->get_query();
+
+    auto it = subs->find(query);
+
+    if (it != subs->end()) {
```

Comment:
This part seems to be common for a number of `find` methods. Can we create a common function?

Reply:
I'm not 100% sure it's worthwhile as this exact code is only reused in two places: https://github.com/realm/realm-js/blob/cf367f2c55f4cc97e7341dd52de4342836d0b8ee/src/js_subscriptions.hpp#L362-L370 and https://github.com/realm/realm-js/blob/cf367f2c55f4cc97e7341dd52de4342836d0b8ee/src/js_subscriptions.hpp#L396-L404, and in fact is calling two different implementations of `find`: https://github.com/realm/realm-core/blob/e0facc4dcbeff538ae1ed5d0779f1613a0cee499/src/realm/sync/subscriptions.hpp#L169-L170 so it would just be `if/else` chunk we could reuse. I am happy to change if you think it is worthwhile though